diff --git a/practica8/src/main/java/mx/unam/ciencias/icc/Arreglos.java b/practica8/src/main/java/mx/unam/ciencias/icc/Arreglos.java
index cb53134..adefac0 100644
--- a/practica8/src/main/java/mx/unam/ciencias/icc/Arreglos.java
+++ b/practica8/src/main/java/mx/unam/ciencias/icc/Arreglos.java
@@ -10,6 +10,20 @@ public class Arreglos {
     /* Constructor privado para evitar instanciación. */
     private Arreglos() {}
 
+    /**
+     * Metodo auxiliar.
+     * Intercambia de posicion los dos elementos del arreglo recibido.
+     * @param <T> tipo del que puede ser el arreglo.
+     * @param arreglo el arreglo a utiliza.
+     * @param i el indice del primer elemento.
+     * @param j el indice del segundo elemento.
+     */
+    private static <T> void swap(T[] arreglo, int i, int j) {
+        T aux = arreglo[i];
+        arreglo[i] = arreglo[j];
+        arreglo[j] = aux;
+    }
+
     /**
      * Ordena el arreglo recibido usando SelectionSort.
      * @param <T> tipo del que puede ser el arreglo.
@@ -28,7 +42,15 @@ public class Arreglos {
      */
     public static <T> void
     selectionSort(T[] arreglo, Comparator<T> comparador) {
-        // Aquí va su código.
+        for (int i = 0; i < arreglo.length - 1; i++) {
+            int min_idx = i;
+            for (int j = i + 1; j < arreglo.length; j++) {
+                if (comparador.compare(arreglo[j], arreglo[min_idx]) < 0)
+                    min_idx = j;
+            }
+            if (min_idx != i)
+                swap(arreglo, i, min_idx);
+        }
     }
 
     /**
@@ -49,7 +71,52 @@ public class Arreglos {
      */
     public static <T> void
     quickSort(T[] arreglo, Comparator<T> comparador) {
-        // Aquí va su código.
+        quickSort(arreglo, comparador, 0, arreglo.length - 1);
+    }
+
+    /**
+     * Ordena el arreglo recibido usando QuickSort.
+     * @param <T> tipo del que puede ser el arreglo.
+     * @param arreglo el arreglo a ordenar.
+     * @param comparador el comparador para ordenar el arreglo.
+     * @param bajo el limite inferior del subarreglo a ordenar.
+     * @param alto el limite superior de subarreglo a ordenar.
+     */
+    private static <T> void 
+    quickSort(T[] arreglo, Comparator<T> comparador, int bajo, int alto) {
+        // Caso Base:
+        if (bajo >= alto)
+            return;
+        // Ponemos al pivote en su posicion ordenada y particionamos el arreglo [menores, pivote, mayore]
+        int pivoteIdx = particionar(arreglo, comparador, bajo, alto);
+        // Llamada recursiva en el subarreglo izquierdo:
+        quickSort(arreglo, comparador, bajo, pivoteIdx - 1);
+        // Llamada recursiva en el subarreglo derecho:
+        quickSort(arreglo, comparador, pivoteIdx + 1, alto);
+    }
+
+    /**
+     * Metodo auxiliar del metodo quickSort.
+     * Particiona el arreglo en [menores, pivote, mayores].
+     * @param <T> tipo del que puede ser el arreglo.
+     * @param arreglo el arreglo a particionar.
+     * @param comparador el comparador para hacer la particion.
+     * @param bajo el limite inferior del subarreglo a particionar.
+     * @param alto el limite superior de subarreglo a ordenar.
+     * @return el indice del arreglo que corresponde a la posicion correcto del pivote.
+     */
+    private static <T> int
+    particionar(T[] arreglo, Comparator<T> comparador, int bajo, int alto) {
+        T pivote = arreglo[alto];
+        int pivoteIdx = bajo - 1;
+        for (int i = bajo; i < alto; i++) {
+            if (comparador.compare(arreglo[i], pivote) <= 0) {
+                pivoteIdx++;
+                swap(arreglo, i, pivoteIdx);
+            }
+        }
+        swap(arreglo, pivoteIdx + 1, alto);
+        return pivoteIdx + 1;
     }
 
     /**
@@ -76,6 +143,35 @@ public class Arreglos {
      */
     public static <T> int
     busquedaBinaria(T[] arreglo, T elemento, Comparator<T> comparador) {
-        // Aquí va su código.
+        return busquedaBinaria(arreglo, elemento, comparador, 0, arreglo.length);
+    }
+
+    /**
+     * Hace una búsqueda binaria del elemento en el arreglo de manera recursiva. 
+     * Regresa el índice del elemento en el arreglo, o -1 si no se encuentra.
+     * @param <T> tipo del que puede ser el arreglo.
+     * @param arreglo el arreglo dónde buscar.
+     * @param elemento el elemento a buscar.
+     * @param comparador el comparador para hacer la búsqueda.
+     * @param inicio la cota inferior de la busqueda.
+     * @param fin la cota superior de la busqueda.
+     * @return el índice del elemento en el arreglo, o -1 si no se encuentra.
+     */
+    private static <T> int
+    busquedaBinaria(T[] arreglo, T elemento, Comparator<T> comparador, int inicio, int fin) {
+        // Caso Base:
+        if (inicio >= fin)
+            return -1;
+        // Elemento medio:
+        int medio = inicio + (fin - inicio) / 2;
+        int comp = comparador.compare(arreglo[medio], elemento);
+        // Busqueda en el subarreglo izquierdo:
+        if (comp > 0)
+            return busquedaBinaria(arreglo, elemento, comparador, inicio, medio);
+        // Busqueda en el suarreglo derecho:
+        if (comp < 0)
+            return busquedaBinaria(arreglo, elemento, comparador, medio + 1, fin);
+        // Se encontro el elemento:
+        return medio;
     }
 }
diff --git a/practica8/src/main/java/mx/unam/ciencias/icc/BaseDeDatos.java b/practica8/src/main/java/mx/unam/ciencias/icc/BaseDeDatos.java
index ed50626..742ef7d 100644
--- a/practica8/src/main/java/mx/unam/ciencias/icc/BaseDeDatos.java
+++ b/practica8/src/main/java/mx/unam/ciencias/icc/BaseDeDatos.java
@@ -3,6 +3,8 @@ package mx.unam.ciencias.icc;
 import java.io.BufferedReader;
 import java.io.BufferedWriter;
 import java.io.IOException;
+import java.util.Iterator;
+import java.util.Iterator;
 
 /**
  * Clase abstracta para bases de datos genéricas. Provee métodos para agregar y
@@ -27,7 +29,7 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * Constructor único.
      */
     public BaseDeDatos() {
-        // Aquí va su código.
+        registros = new Lista<>();
     }
 
     /**
@@ -35,7 +37,7 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * @return el número de registros en la base de datos.
      */
     public int getNumRegistros() {
-        // Aquí va su código.
+        return registros.getLongitud();
     }
 
     /**
@@ -44,7 +46,7 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * @return una lista con los registros en la base de datos.
      */
     public Lista<R> getRegistros() {
-        // Aquí va su código.
+        return registros.copia();
     }
 
     /**
@@ -52,7 +54,7 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * @param registro el registro que hay que agregar a la base de datos.
      */
     public void agregaRegistro(R registro) {
-        // Aquí va su código.
+        registros.agregaFinal(registro);
     }
 
     /**
@@ -60,14 +62,14 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * @param registro el registro que hay que eliminar de la base de datos.
      */
     public void eliminaRegistro(R registro) {
-        // Aquí va su código.
+        registros.elimina(registro);
     }
 
     /**
      * Limpia la base de datos.
      */
     public void limpia() {
-        // Aquí va su código.
+        registros.limpia();
     }
 
     /**
@@ -76,7 +78,14 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * @throws IOException si ocurre un error de entrada/salida.
      */
     public void guarda(BufferedWriter out) throws IOException {
-        // Aquí va su código.
+        Iterator<R> iterator = registros.iterator();
+        while (iterator.hasNext()) {
+            try {
+                out.append(iterator.next().seria());
+            } catch (IOException io) {
+                throw new IOException("Error al escribir los registros en el archivo.");
+            }
+        }
     }
 
     /**
@@ -87,7 +96,17 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * @throws IOException si ocurre un error de entrada/salida.
      */
     public void carga(BufferedReader in) throws IOException {
-        // Aquí va su código.
+        limpia();
+        String linea = "";
+        while ((linea = in.readLine()) != null) {
+            R registro = creaRegistro();
+            try {
+                registro.deseria(linea);
+            } catch (ExcepcionLineaInvalida eli) {
+                return;
+            }
+            registros.agregaFinal(registro);
+        }
     }
 
     /**
@@ -100,7 +119,18 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      *         correcta.
      */
     public Lista<R> buscaRegistros(C campo, Object valor) {
-        // Aquí va su código.
+        if (!(campo instanceof Enum))
+            throw new IllegalArgumentException("El campo recibido no es una Enumeracion.");
+        Lista<R> casa = new Lista<>();
+        if (valor == null)
+            return casa;
+        Iterator<R> iterator = registros.iterator();
+        while (iterator.hasNext()) {
+            R registo = iterator.next();
+            if (registo.casa(campo, valor))
+                casa.agregaFinal(registo);
+        }
+        return casa;
     }
 
     /**
diff --git a/practica8/src/main/java/mx/unam/ciencias/icc/BaseDeDatosEstudiantes.java b/practica8/src/main/java/mx/unam/ciencias/icc/BaseDeDatosEstudiantes.java
index 99253e8..2aab6cc 100644
--- a/practica8/src/main/java/mx/unam/ciencias/icc/BaseDeDatosEstudiantes.java
+++ b/practica8/src/main/java/mx/unam/ciencias/icc/BaseDeDatosEstudiantes.java
@@ -11,6 +11,6 @@ public class BaseDeDatosEstudiantes
      * @return un estudiante en blanco.
      */
     @Override public Estudiante creaRegistro() {
-        // Aquí va su código.
+        return new Estudiante(null, 0, 0.0, 0);
     }
 }
diff --git a/practica8/src/main/java/mx/unam/ciencias/icc/CampoEstudiante.java b/practica8/src/main/java/mx/unam/ciencias/icc/CampoEstudiante.java
index 4a8740d..a6d0246 100644
--- a/practica8/src/main/java/mx/unam/ciencias/icc/CampoEstudiante.java
+++ b/practica8/src/main/java/mx/unam/ciencias/icc/CampoEstudiante.java
@@ -20,6 +20,12 @@ public enum CampoEstudiante {
      * @return una representación en cadena del campo.
      */
     @Override public String toString() {
-        // Aquí va su código.
+        switch (this) {
+            case NOMBRE: return "Nombre";
+            case CUENTA: return "# Cuenta";
+            case PROMEDIO: return "Promedio";
+            case EDAD: return "Edad";
+            default: throw new IllegalArgumentException("Error en el toString() de CampoEstudiante.");
+        }
     }
 }
diff --git a/practica8/src/main/java/mx/unam/ciencias/icc/Estudiante.java b/practica8/src/main/java/mx/unam/ciencias/icc/Estudiante.java
index a8b7da3..e1be9ca 100644
--- a/practica8/src/main/java/mx/unam/ciencias/icc/Estudiante.java
+++ b/practica8/src/main/java/mx/unam/ciencias/icc/Estudiante.java
@@ -29,7 +29,10 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
                       int    cuenta,
                       double promedio,
                       int    edad) {
-        // Aquí va su código.
+        this.nombre = nombre;
+        this.cuenta = cuenta;
+        this.promedio = promedio;
+        this.edad = edad;
     }
 
     /**
@@ -37,7 +40,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @return el nombre del estudiante.
      */
     public String getNombre() {
-        // Aquí va su código.
+        return nombre;
     }
 
     /**
@@ -45,7 +48,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @param nombre el nuevo nombre del estudiante.
      */
     public void setNombre(String nombre) {
-        // Aquí va su código.
+        this.nombre = nombre;
     }
 
     /**
@@ -53,7 +56,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @return el número de cuenta del estudiante.
      */
     public int getCuenta() {
-        // Aquí va su código.
+        return cuenta;
     }
 
     /**
@@ -61,7 +64,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @param cuenta el nuevo número de cuenta del estudiante.
      */
     public void setCuenta(int cuenta) {
-        // Aquí va su código.
+        this.cuenta = cuenta;
     }
 
     /**
@@ -69,7 +72,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @return el promedio del estudiante.
      */
     public double getPromedio() {
-        // Aquí va su código.
+        return promedio;
     }
 
     /**
@@ -77,7 +80,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @param promedio el nuevo promedio del estudiante.
      */
     public void setPromedio(double promedio) {
-        // Aquí va su código.
+        this.promedio = promedio;
     }
 
     /**
@@ -85,7 +88,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @return la edad del estudiante.
      */
     public int getEdad() {
-        // Aquí va su código.
+        return edad;
     }
 
     /**
@@ -93,7 +96,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @param edad la nueva edad del estudiante.
      */
     public void setEdad(int edad) {
-        // Aquí va su código.
+        this.edad = edad;
     }
 
     /**
@@ -101,7 +104,11 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @return una representación en cadena del estudiante.
      */
     @Override public String toString() {
-        // Aquí va su código.
+        return String.format("Nombre   : %s\n" +
+                             "Cuenta   : %09d\n" + 
+                             "Promedio : %2.2f\n" + 
+                             "Edad     : %d", 
+                             nombre, cuenta, promedio, edad);
     }
 
     /**
@@ -116,7 +123,10 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
         if (!(objeto instanceof Estudiante))
             return false;
         Estudiante estudiante = (Estudiante)objeto;
-        // Aquí va su código.
+        return nombre.equals(estudiante.nombre) &&
+                cuenta == estudiante.cuenta &&
+                promedio == estudiante.promedio &&
+                edad == estudiante.edad;
     }
 
     /**
@@ -126,7 +136,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @return la seriación del estudiante en una línea de texto.
      */
     @Override public String seria() {
-        // Aquí va su código.
+        return String.format("%s\t%d\t%2.2f\t%d\n", nombre, cuenta, promedio, edad);
     }
 
     /**
@@ -138,7 +148,20 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      *         es una seriación válida de un estudiante.
      */
     @Override public void deseria(String linea) {
-        // Aquí va su código.
+        String MSJ_ERROR = "la línea recibida no es una seriación válida de un estudiante.";
+        if (linea == null || linea.isEmpty())
+            throw new ExcepcionLineaInvalida(MSJ_ERROR);
+        String[] datos = linea.trim().split("\t");
+        if (datos.length != 4)
+            throw new ExcepcionLineaInvalida(MSJ_ERROR);
+        nombre = datos[0];
+        try {
+            cuenta = Integer.parseInt(datos[1]);
+            promedio = Double.parseDouble(datos[2]);
+            edad = Integer.parseInt(datos[3]);
+        } catch (NumberFormatException nfe) {
+            throw new ExcepcionLineaInvalida(MSJ_ERROR);
+        }
     }
 
     /**
@@ -147,7 +170,12 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @throws IllegalArgumentException si el estudiante es <code>null</code>.
      */
     @Override public void actualiza(Estudiante estudiante) {
-        // Aquí va su código.
+        if (estudiante == null)
+            throw new IllegalArgumentException("El objeto estudiante recibido es null.");
+        nombre = estudiante.nombre;
+        cuenta = estudiante.cuenta;
+        promedio = estudiante.promedio;
+        edad = estudiante.edad;
     }
 
     /**
@@ -176,6 +204,30 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @throws IllegalArgumentException si el campo es <code>null</code>.
      */
     @Override public boolean casa(CampoEstudiante campo, Object valor) {
-        // Aquí va su código.
+        if (campo == null)
+            throw new IllegalArgumentException("El campo recibido es null.");
+        if (valor == null)
+            return false;
+        switch (campo) {
+            case NOMBRE:
+                if (valor instanceof String) {
+                    String val = (String) valor;
+                    return !val.isEmpty() && nombre.indexOf(val) != -1;
+                }
+                return false;
+            case CUENTA: 
+                if (valor instanceof Integer) 
+                    return (Integer) valor <= cuenta;
+                return false;
+            case PROMEDIO:
+                if (valor instanceof Double) 
+                    return (Double) valor <= promedio;
+                return false;
+            case EDAD:
+                if (valor instanceof Integer)
+                    return (Integer) valor <= edad;
+            default: return false;
+                //throw new IllegalArgumentException("Error en el CampoEstudiante recibido.");
+        }
     }
 }
diff --git a/practica8/src/main/java/mx/unam/ciencias/icc/Lista.java b/practica8/src/main/java/mx/unam/ciencias/icc/Lista.java
index abc8099..b8be5cc 100644
--- a/practica8/src/main/java/mx/unam/ciencias/icc/Lista.java
+++ b/practica8/src/main/java/mx/unam/ciencias/icc/Lista.java
@@ -1,5 +1,7 @@
 package mx.unam.ciencias.icc;
 
+import java.util.Iterator;
+import java.util.Iterator;
 import java.util.Comparator;
 import java.util.Iterator;
 import java.util.NoSuchElementException;
@@ -30,7 +32,7 @@ public class Lista<T> implements Iterable<T> {
 
         /* Construye un nodo con un elemento. */
         private Nodo(T elemento) {
-            // Aquí va su código.
+            this.elemento = elemento;
         }
     }
 
@@ -43,37 +45,47 @@ public class Lista<T> implements Iterable<T> {
 
         /* Construye un nuevo iterador. */
         private Iterador() {
-            // Aquí va su código.
+            start();
         }
 
         /* Nos dice si hay un elemento siguiente. */
         @Override public boolean hasNext() {
-            // Aquí va su código.
+            return siguiente != null;
         }
 
         /* Nos da el elemento siguiente. */
         @Override public T next() {
-            // Aquí va su código.
+            if (!hasNext())
+                throw new NoSuchElementException("No hay elemento siguiente.");
+            anterior = siguiente;
+            siguiente = siguiente.siguiente;
+            return anterior.elemento;
         }
 
         /* Nos dice si hay un elemento anterior. */
         @Override public boolean hasPrevious() {
-            // Aquí va su código.
+            return anterior != null;
         }
 
         /* Nos da el elemento anterior. */
         @Override public T previous() {
-            // Aquí va su código.
+            if (!hasPrevious())
+                throw new NoSuchElementException("No hay elemento anterior.");
+            siguiente = anterior;
+            anterior = anterior.anterior;
+            return siguiente.elemento;
         }
 
         /* Mueve el iterador al inicio de la lista. */
         @Override public void start() {
-            // Aquí va su código.
+            anterior = null;
+            siguiente = cabeza;
         }
 
         /* Mueve el iterador al final de la lista. */
         @Override public void end() {
-            // Aquí va su código.
+            anterior = rabo;
+            siguiente = null;
         }
     }
 
@@ -83,13 +95,17 @@ public class Lista<T> implements Iterable<T> {
     private Nodo rabo;
     /* Número de elementos en la lista. */
     private int longitud;
+    /* Mensaje de error al tratar de agregar un elemento null a la lista. */
+    private static final String MSJ_NULL_ELEM = "La lista no acepta a null como elemento.";
+    /* Mensaje de error al tratar de eliminar cuando la lista esta vacia. */
+    private static final String MSJ_EMPTY_LIST = "La lista esta vacio, no hay elementos.";
 
     /**
      * Regresa la longitud de la lista.
      * @return la longitud de la lista, el número de elementos que contiene.
      */
     public int getLongitud() {
-        // Aquí va su código.
+        return longitud;
     }
 
     /**
@@ -98,7 +114,7 @@ public class Lista<T> implements Iterable<T> {
      *         otro caso.
      */
     public boolean esVacia() {
-        // Aquí va su código.
+        return cabeza == null;
     }
 
     /**
@@ -109,7 +125,17 @@ public class Lista<T> implements Iterable<T> {
      *         <code>null</code>.
      */
     public void agregaFinal(T elemento) {
-        // Aquí va su código.
+        if (elemento == null)
+            throw new IllegalArgumentException(MSJ_NULL_ELEM);
+        Nodo nuevo = new Nodo(elemento);
+        if (esVacia()) {
+            cabeza = rabo = nuevo;
+        } else {
+            rabo.siguiente = nuevo;
+            nuevo.anterior = rabo;
+            rabo = nuevo;
+        }
+        longitud++;
     }
 
     /**
@@ -120,7 +146,17 @@ public class Lista<T> implements Iterable<T> {
      *         <code>null</code>.
      */
     public void agregaInicio(T elemento) {
-        // Aquí va su código.
+        if (elemento == null)
+            throw new IllegalArgumentException(MSJ_NULL_ELEM);
+        Nodo nuevo = new Nodo(elemento);
+        if (esVacia()) {
+            cabeza = rabo = nuevo;
+        } else {
+            cabeza.anterior = nuevo;
+            nuevo.siguiente = cabeza;
+            cabeza = nuevo;
+        }
+        longitud++;
     }
 
     /**
@@ -139,7 +175,26 @@ public class Lista<T> implements Iterable<T> {
      *         <code>null</code>.
      */
     public void inserta(int i, T elemento) {
-        // Aquí va su código.
+        if (elemento == null)
+            throw new IllegalArgumentException(MSJ_NULL_ELEM);
+        if (i <= 0) {
+            agregaInicio(elemento);
+            return;
+        }
+        if (i >= longitud) {
+            agregaFinal(elemento);
+            return;
+        }
+        Nodo temp = cabeza;
+        for (int j = 0; j < i - 1; j++) {
+            temp = temp.siguiente;
+        }
+        Nodo nuevo = new Nodo(elemento);
+        nuevo.anterior = temp;
+        nuevo.siguiente = temp.siguiente;
+        temp.siguiente.anterior = nuevo;
+        temp.siguiente = nuevo;
+        longitud++;
     }
 
     /**
@@ -148,7 +203,41 @@ public class Lista<T> implements Iterable<T> {
      * @param elemento el elemento a eliminar.
      */
     public void elimina(T elemento) {
-        // Aquí va su código.
+        if (esVacia() || elemento == null)
+            return;
+        Nodo elim = getNodo(elemento);
+        if (elim == null)
+            return;
+        if (elim == cabeza) {
+            eliminaPrimero();
+            return;
+        }
+        if (elim == rabo) {
+            eliminaUltimo();
+            return;
+        }
+        elim.anterior.siguiente = elim.siguiente;
+        elim.siguiente.anterior = elim.anterior;
+        longitud--;
+    }
+
+    /**
+     * Metodo auxiliar al metodo elimina.
+     * Retorna el Nodo que contega la primer ocurrencia del elemento recibido.
+     * @param elemento el elemento a buscar el la lista.
+     * @return el Nodo que contega al elemento, o <code>null</code> si el elemento 
+     * no esta en la lista.
+     */
+    private Nodo getNodo(T elemento) {
+        if (esVacia() || elemento == null) 
+            return null;
+        Nodo temp = cabeza;
+        while (temp != null) {
+            if (temp.elemento.equals(elemento))
+                return temp;
+            temp = temp.siguiente;
+        }
+        return null;
     }
 
     /**
@@ -157,7 +246,17 @@ public class Lista<T> implements Iterable<T> {
      * @throws NoSuchElementException si la lista es vacía.
      */
     public T eliminaPrimero() {
-        // Aquí va su código.
+        if (esVacia())
+            throw new NoSuchElementException(MSJ_EMPTY_LIST);
+        T elim = cabeza.elemento;
+        if (cabeza == rabo) {
+            cabeza = rabo = null;
+        } else {
+            cabeza = cabeza.siguiente;
+            cabeza.anterior = null;
+        }
+        longitud--;
+        return elim;
     }
 
     /**
@@ -166,7 +265,17 @@ public class Lista<T> implements Iterable<T> {
      * @throws NoSuchElementException si la lista es vacía.
      */
     public T eliminaUltimo() {
-        // Aquí va su código.
+        if (esVacia())
+            throw new NoSuchElementException(MSJ_EMPTY_LIST);
+        T elim = rabo.elemento;
+        if (cabeza == rabo) {
+            cabeza = rabo = null;
+        } else {
+            rabo = rabo.anterior;
+            rabo.siguiente = null;
+        }
+        longitud--;
+        return elim;
     }
 
     /**
@@ -176,7 +285,7 @@ public class Lista<T> implements Iterable<T> {
      *         <code>false</code> en otro caso.
      */
     public boolean contiene(T elemento) {
-        // Aquí va su código.
+        return getNodo(elemento) != null;
     }
 
     /**
@@ -184,7 +293,12 @@ public class Lista<T> implements Iterable<T> {
      * @return una nueva lista que es la reversa la que manda llamar el método.
      */
     public Lista<T> reversa() {
-        // Aquí va su código.
+        Lista<T> reversa = new Lista<>();
+        Iterator<T> iterator = iterator();
+        while (iterator.hasNext()) {
+            reversa.agregaInicio(iterator.next());
+        }
+        return reversa;
     }
 
     /**
@@ -193,14 +307,20 @@ public class Lista<T> implements Iterable<T> {
      * @return una copiad de la lista.
      */
     public Lista<T> copia() {
-        // Aquí va su código.
+        Lista<T> copia = new Lista<>();
+        Iterator<T> iterator = iterator();
+        while (iterator.hasNext()) {
+            copia.agregaFinal(iterator.next());
+        }
+        return copia;
     }
 
     /**
      * Limpia la lista de elementos, dejándola vacía.
      */
     public void limpia() {
-        // Aquí va su código.
+        cabeza = rabo = null;
+        longitud = 0;
     }
 
     /**
@@ -209,7 +329,9 @@ public class Lista<T> implements Iterable<T> {
      * @throws NoSuchElementException si la lista es vacía.
      */
     public T getPrimero() {
-        // Aquí va su código.
+        if (esVacia())
+            throw new NoSuchElementException(MSJ_EMPTY_LIST);
+        return cabeza.elemento;
     }
 
     /**
@@ -218,7 +340,9 @@ public class Lista<T> implements Iterable<T> {
      * @throws NoSuchElementException si la lista es vacía.
      */
     public T getUltimo() {
-        // Aquí va su código.
+        if (esVacia())
+            throw new NoSuchElementException(MSJ_EMPTY_LIST);
+        return rabo.elemento;
     }
 
     /**
@@ -229,7 +353,13 @@ public class Lista<T> implements Iterable<T> {
      *         igual que el número de elementos en la lista.
      */
     public T get(int i) {
-        // Aquí va su código.
+        if (i < 0 || i >= longitud)
+            throw new ExcepcionIndiceInvalido("Indice fuera de rango: [0, longitud).");
+        Nodo temp = cabeza;
+        for (int j = 0; j < i; j++) {
+            temp = temp.siguiente;
+        }
+        return temp.elemento;
     }
 
     /**
@@ -239,7 +369,15 @@ public class Lista<T> implements Iterable<T> {
      *         no está contenido en la lista.
      */
     public int indiceDe(T elemento) {
-        // Aquí va su código.
+        if (esVacia() || elemento == null)
+            return -1;
+        Nodo temp = cabeza;
+        for (int i = 0; i < longitud; i++) {
+            if (temp.elemento.equals(elemento))
+                return i;
+            temp = temp.siguiente;
+        }
+        return -1;
     }
 
     /**
@@ -247,7 +385,14 @@ public class Lista<T> implements Iterable<T> {
      * @return una representación en cadena de la lista.
      */
     @Override public String toString() {
-        // Aquí va su código.
+        StringBuilder sb = new StringBuilder("[");
+        Iterator<T> iterator = iterator();
+        while (iterator.hasNext()) {
+            sb.append(iterator.next());
+            if (iterator.hasNext())
+                sb.append(", ");
+        }
+        return sb.toString() + "]";
     }
 
     /**
@@ -260,7 +405,15 @@ public class Lista<T> implements Iterable<T> {
         if (objeto == null || getClass() != objeto.getClass())
             return false;
         @SuppressWarnings("unchecked") Lista<T> lista = (Lista<T>)objeto;
-        // Aquí va su código.
+        if (longitud != lista.longitud)
+            return false;
+        Iterator<T> ite1 = iterator();
+        Iterator<T> ite2 = lista.iterator();
+        for (int i = 0; i < longitud; i++) {
+            if (!ite1.next().equals(ite2.next()))
+                return false;
+        }
+        return true;
     }
 
     /**
@@ -288,7 +441,57 @@ public class Lista<T> implements Iterable<T> {
      * @return una copia de la lista, pero ordenada.
      */
     public Lista<T> mergeSort(Comparator<T> comparador) {
-        // Aquí va su código.
+        // Case Base:
+        if (longitud <= 1)
+            return this.copia();
+        // Dividir la lista en dos mitades:
+        Lista<T> lista1 = new Lista<>();
+        Lista<T> lista2 = new Lista<>();
+        int mitad = longitud / 2;
+        Nodo temp = cabeza;
+        for (int i = 0; i < longitud; i++) {
+            if (i < mitad) {
+                lista1.agregaFinal(temp.elemento);
+            } else {
+                lista2.agregaFinal(temp.elemento);
+            }
+            temp = temp.siguiente;
+        }
+        // Ordenar recursivamente cada mitad:
+        lista1 = lista1.mergeSort(comparador);
+        lista2 = lista2.mergeSort(comparador);
+        // Mezclar las dos mitades ordenadas:
+        return mezcla(comparador, lista1, lista2);
+    }
+
+    /**
+     * Metodo auxiliar al metodo mergeSort.
+     * Mezcla las dos listas recibinas en una nueva lista ordenada.
+     * @param comparador el comparador que lse usará para hacer el ordenamiento.
+     * @param lista1 la primer lista ordenada.
+     * @param lista2 la segunda lista ordenada.
+     * @return una nueva lista ordenada con los elementos de lista1 y lista2. 
+     */
+    private Lista<T> mezcla(Comparator<T> comparador, Lista<T> lista1, Lista<T> lista2) {
+        Lista<T> mezcla = new Lista<>();
+        Nodo n1 = lista1.cabeza;
+        Nodo n2 = lista2.cabeza;
+        while (n1 != null && n2 != null) {
+            int comp = comparador.compare(n1.elemento, n2.elemento);
+            if (comp <= 0) {
+                mezcla.agregaFinal(n1.elemento);
+                n1 = n1.siguiente;
+            } else {
+                mezcla.agregaFinal(n2.elemento);
+                n2 = n2.siguiente;
+            }
+        }
+        n1 = (n1 != null) ? n1 : n2;
+        while (n1 != null) {
+            mezcla.agregaFinal(n1.elemento);
+            n1 = n1.siguiente;
+        }
+        return mezcla;
     }
 
     /**
@@ -313,7 +516,18 @@ public class Lista<T> implements Iterable<T> {
      *         <code>false</code> en otro caso.
      */
     public boolean busquedaLineal(T elemento, Comparator<T> comparador) {
-        // Aquí va su código.
+        if (esVacia() || elemento == null)
+            return false;
+        Nodo temp = cabeza;
+        while (temp != null) {
+            int comp = comparador.compare(temp.elemento, elemento);
+            if (comp == 0)
+                return true;
+            if (comp > 1)
+                return false;
+            temp = temp.siguiente;
+        }
+        return false;
     }
 
     /**
