diff --git a/practica7/src/main/java/mx/unam/ciencias/icc/BaseDeDatos.java b/practica7/src/main/java/mx/unam/ciencias/icc/BaseDeDatos.java
index ed50626..93a5264 100644
--- a/practica7/src/main/java/mx/unam/ciencias/icc/BaseDeDatos.java
+++ b/practica7/src/main/java/mx/unam/ciencias/icc/BaseDeDatos.java
@@ -3,6 +3,7 @@ package mx.unam.ciencias.icc;
 import java.io.BufferedReader;
 import java.io.BufferedWriter;
 import java.io.IOException;
+import java.util.Iterator;
 
 /**
  * Clase abstracta para bases de datos genéricas. Provee métodos para agregar y
@@ -27,7 +28,7 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * Constructor único.
      */
     public BaseDeDatos() {
-        // Aquí va su código.
+        registros = new Lista<>();
     }
 
     /**
@@ -35,7 +36,7 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * @return el número de registros en la base de datos.
      */
     public int getNumRegistros() {
-        // Aquí va su código.
+        return registros.getLongitud();
     }
 
     /**
@@ -44,7 +45,7 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * @return una lista con los registros en la base de datos.
      */
     public Lista<R> getRegistros() {
-        // Aquí va su código.
+        return registros.copia();
     }
 
     /**
@@ -52,7 +53,7 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * @param registro el registro que hay que agregar a la base de datos.
      */
     public void agregaRegistro(R registro) {
-        // Aquí va su código.
+        registros.agregaFinal(registro);;
     }
 
     /**
@@ -60,14 +61,14 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * @param registro el registro que hay que eliminar de la base de datos.
      */
     public void eliminaRegistro(R registro) {
-        // Aquí va su código.
+        registros.elimina(registro);
     }
 
     /**
      * Limpia la base de datos.
      */
     public void limpia() {
-        // Aquí va su código.
+        registros.limpia();
     }
 
     /**
@@ -76,7 +77,10 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * @throws IOException si ocurre un error de entrada/salida.
      */
     public void guarda(BufferedWriter out) throws IOException {
-        // Aquí va su código.
+        Iterator<R> it = registros.iterator();
+        while (it.hasNext()) {
+            out.append(it.next().seria());
+        }
     }
 
     /**
@@ -87,7 +91,17 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      * @throws IOException si ocurre un error de entrada/salida.
      */
     public void carga(BufferedReader in) throws IOException {
-        // Aquí va su código.
+        limpia();
+        String linea = "";
+        while ((linea = in.readLine()) != null) {
+            R r = creaRegistro();
+            try {
+                r.deseria(linea);
+            } catch (ExcepcionLineaInvalida eli) {
+                return;
+            }
+            registros.agregaFinal(r);
+        }
     }
 
     /**
@@ -100,7 +114,16 @@ public abstract class BaseDeDatos<R extends Registro<R, C>, C extends Enum> {
      *         correcta.
      */
     public Lista<R> buscaRegistros(C campo, Object valor) {
-        // Aquí va su código.
+        if (campo == null || !(campo instanceof Enum))
+            throw new IllegalArgumentException("El campo recibido no es instancia de C");
+        Lista<R> casa = new Lista<>();
+        Iterator<R> it = registros.iterator();
+        while (it.hasNext()) {
+            R r = it.next();
+            if (r.casa(campo, valor))
+                casa.agregaFinal(r);
+        }
+        return casa;
     }
 
     /**
diff --git a/practica7/src/main/java/mx/unam/ciencias/icc/BaseDeDatosEstudiantes.java b/practica7/src/main/java/mx/unam/ciencias/icc/BaseDeDatosEstudiantes.java
index 99253e8..2aab6cc 100644
--- a/practica7/src/main/java/mx/unam/ciencias/icc/BaseDeDatosEstudiantes.java
+++ b/practica7/src/main/java/mx/unam/ciencias/icc/BaseDeDatosEstudiantes.java
@@ -11,6 +11,6 @@ public class BaseDeDatosEstudiantes
      * @return un estudiante en blanco.
      */
     @Override public Estudiante creaRegistro() {
-        // Aquí va su código.
+        return new Estudiante(null, 0, 0.0, 0);
     }
 }
diff --git a/practica7/src/main/java/mx/unam/ciencias/icc/CampoEstudiante.java b/practica7/src/main/java/mx/unam/ciencias/icc/CampoEstudiante.java
index 4a8740d..4269088 100644
--- a/practica7/src/main/java/mx/unam/ciencias/icc/CampoEstudiante.java
+++ b/practica7/src/main/java/mx/unam/ciencias/icc/CampoEstudiante.java
@@ -20,6 +20,13 @@ public enum CampoEstudiante {
      * @return una representación en cadena del campo.
      */
     @Override public String toString() {
-        // Aquí va su código.
+        switch (this) {
+            case NOMBRE: return "Nombre";
+            case CUENTA: return "# Cuenta";
+            case PROMEDIO: return "Promedio";
+            case EDAD: return "Edad";
+            default:
+                throw new IllegalArgumentException("Error en el toString() de CampoEstudiante.");
+        }
     }
 }
diff --git a/practica7/src/main/java/mx/unam/ciencias/icc/Estudiante.java b/practica7/src/main/java/mx/unam/ciencias/icc/Estudiante.java
index a8b7da3..e022c58 100644
--- a/practica7/src/main/java/mx/unam/ciencias/icc/Estudiante.java
+++ b/practica7/src/main/java/mx/unam/ciencias/icc/Estudiante.java
@@ -29,7 +29,10 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
                       int    cuenta,
                       double promedio,
                       int    edad) {
-        // Aquí va su código.
+        this.nombre = nombre;
+        this.cuenta = cuenta;
+        this.promedio = promedio;
+        this.edad = edad;
     }
 
     /**
@@ -37,7 +40,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @return el nombre del estudiante.
      */
     public String getNombre() {
-        // Aquí va su código.
+        return nombre;
     }
 
     /**
@@ -45,7 +48,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @param nombre el nuevo nombre del estudiante.
      */
     public void setNombre(String nombre) {
-        // Aquí va su código.
+        this.nombre = nombre;
     }
 
     /**
@@ -53,7 +56,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @return el número de cuenta del estudiante.
      */
     public int getCuenta() {
-        // Aquí va su código.
+        return cuenta;
     }
 
     /**
@@ -61,7 +64,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @param cuenta el nuevo número de cuenta del estudiante.
      */
     public void setCuenta(int cuenta) {
-        // Aquí va su código.
+        this.cuenta = cuenta;
     }
 
     /**
@@ -69,7 +72,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @return el promedio del estudiante.
      */
     public double getPromedio() {
-        // Aquí va su código.
+        return promedio;
     }
 
     /**
@@ -77,7 +80,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @param promedio el nuevo promedio del estudiante.
      */
     public void setPromedio(double promedio) {
-        // Aquí va su código.
+        this.promedio = promedio;
     }
 
     /**
@@ -85,7 +88,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @return la edad del estudiante.
      */
     public int getEdad() {
-        // Aquí va su código.
+        return edad;
     }
 
     /**
@@ -93,7 +96,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @param edad la nueva edad del estudiante.
      */
     public void setEdad(int edad) {
-        // Aquí va su código.
+        this.edad = edad;
     }
 
     /**
@@ -101,7 +104,11 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @return una representación en cadena del estudiante.
      */
     @Override public String toString() {
-        // Aquí va su código.
+        return String.format("Nombre   : %s\n" +
+                             "Cuenta   : %09d\n" +
+                             "Promedio : %2.2f\n" +
+                             "Edad     : %d", 
+                             nombre, cuenta, promedio, edad);
     }
 
     /**
@@ -116,7 +123,10 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
         if (!(objeto instanceof Estudiante))
             return false;
         Estudiante estudiante = (Estudiante)objeto;
-        // Aquí va su código.
+        return nombre.equals(estudiante.nombre) &&
+               cuenta == estudiante.cuenta &&
+               promedio == estudiante.promedio &&
+               edad == estudiante.edad;
     }
 
     /**
@@ -126,7 +136,7 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @return la seriación del estudiante en una línea de texto.
      */
     @Override public String seria() {
-        // Aquí va su código.
+        return String.format("%s\t%d\t%2.2f\t%d\n", nombre, cuenta, promedio, edad);
     }
 
     /**
@@ -138,7 +148,20 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      *         es una seriación válida de un estudiante.
      */
     @Override public void deseria(String linea) {
-        // Aquí va su código.
+        String MSJ_ERROR = "La línea recibida es no es una seriación válida de un estudiante.";
+        if (linea == null || linea.isEmpty())
+            throw new ExcepcionLineaInvalida(MSJ_ERROR);
+        String[] datos = linea.trim().split("\t");
+        if (datos.length != 4)
+            throw new ExcepcionLineaInvalida(MSJ_ERROR);
+        nombre = datos[0];
+        try {
+            cuenta = Integer.parseInt(datos[1]);
+            promedio = Double.parseDouble(datos[2]);
+            edad = Integer.parseInt(datos[3]);
+        } catch (NumberFormatException nfe) {
+            throw new ExcepcionLineaInvalida(MSJ_ERROR);
+        }
     }
 
     /**
@@ -147,7 +170,12 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @throws IllegalArgumentException si el estudiante es <code>null</code>.
      */
     @Override public void actualiza(Estudiante estudiante) {
-        // Aquí va su código.
+        if (estudiante == null)
+            throw new IllegalArgumentException("El estudiante recibido es null.");
+        nombre = estudiante.nombre;
+        cuenta = estudiante.cuenta;
+        promedio = estudiante.promedio;
+        edad = estudiante.edad;
     }
 
     /**
@@ -176,6 +204,34 @@ public class Estudiante implements Registro<Estudiante, CampoEstudiante> {
      * @throws IllegalArgumentException si el campo es <code>null</code>.
      */
     @Override public boolean casa(CampoEstudiante campo, Object valor) {
-        // Aquí va su código.
+        if (campo == null)
+            throw new IllegalArgumentException("El CampoEstudiante recibido es null.");
+        if (valor == null)
+            return false;
+        switch (campo) {
+            case NOMBRE:
+                if (valor instanceof String) {
+                    String val = (String) valor;
+                    return !val.isEmpty() && nombre.indexOf(val) != -1;
+                }
+                return false;
+            case CUENTA:
+                if (valor instanceof Integer) {
+                    return (Integer) valor <= cuenta;
+                }
+                return false;
+            case PROMEDIO:
+                if (valor instanceof Double) {
+                    return (Double) valor <= promedio;  
+                }
+                return false;
+            case EDAD:
+                if (valor instanceof Integer) {
+                    return (Integer) valor <= edad;
+                }
+                return false;
+            default:
+                throw new IllegalArgumentException("El CampoEstudiante no casa con ningun tipo.");
+        }
     }
 }
diff --git a/practica7/src/main/java/mx/unam/ciencias/icc/Lista.java b/practica7/src/main/java/mx/unam/ciencias/icc/Lista.java
index 9ce75fc..dfc1a85 100644
--- a/practica7/src/main/java/mx/unam/ciencias/icc/Lista.java
+++ b/practica7/src/main/java/mx/unam/ciencias/icc/Lista.java
@@ -29,7 +29,7 @@ public class Lista<T> implements Iterable<T> {
 
         /* Construye un nodo con un elemento. */
         private Nodo(T elemento) {
-            // Aquí va su código.
+            this.elemento = elemento;
         }
     }
 
@@ -42,37 +42,47 @@ public class Lista<T> implements Iterable<T> {
 
         /* Construye un nuevo iterador. */
         private Iterador() {
-            // Aquí va su código.
+            start();
         }
 
         /* Nos dice si hay un elemento siguiente. */
         @Override public boolean hasNext() {
-            // Aquí va su código.
+            return siguiente != null;
         }
 
         /* Nos da el elemento siguiente. */
         @Override public T next() {
-            // Aquí va su código.
+            if (!hasNext())
+                throw new NoSuchElementException("No hay elemento siguinte.");
+            anterior = siguiente;
+            siguiente = siguiente.siguiente;
+            return anterior.elemento;
         }
 
         /* Nos dice si hay un elemento anterior. */
         @Override public boolean hasPrevious() {
-            // Aquí va su código.
+            return anterior != null;
         }
 
         /* Nos da el elemento anterior. */
         @Override public T previous() {
-            // Aquí va su código.
+            if (!hasPrevious())
+                throw new NoSuchElementException("No hay elemento anterios.");
+            siguiente = anterior;
+            anterior = anterior.anterior;
+            return siguiente.elemento;
         }
 
         /* Mueve el iterador al inicio de la lista. */
         @Override public void start() {
-            // Aquí va su código.
+            anterior = null;
+            siguiente = cabeza;
         }
 
         /* Mueve el iterador al final de la lista. */
         @Override public void end() {
-            // Aquí va su código.
+            anterior = rabo;
+            siguiente = null;
         }
     }
 
@@ -82,13 +92,17 @@ public class Lista<T> implements Iterable<T> {
     private Nodo rabo;
     /* Número de elementos en la lista. */
     private int longitud;
+    /* Mensaje al tratar agregar un elemento null a la lista. */
+    private static final String MSJ_NULL_ELEM = "La lista no acepta a null como elemento";
+    /* Mensaje al tratar de eliminar cuando la lista esta vacia. */
+    private static final String MSJ_EMPTY_LIST = "La lista esta vacia, no tiene elementos.";
 
     /**
      * Regresa la longitud de la lista.
      * @return la longitud de la lista, el número de elementos que contiene.
      */
     public int getLongitud() {
-        // Aquí va su código.
+        return longitud;
     }
 
     /**
@@ -97,7 +111,7 @@ public class Lista<T> implements Iterable<T> {
      *         otro caso.
      */
     public boolean esVacia() {
-        // Aquí va su código.
+        return cabeza == null;
     }
 
     /**
@@ -108,7 +122,17 @@ public class Lista<T> implements Iterable<T> {
      *         <code>null</code>.
      */
     public void agregaFinal(T elemento) {
-        // Aquí va su código.
+        if (elemento == null) 
+            throw new IllegalArgumentException(MSJ_NULL_ELEM);
+        Nodo nuevo = new Nodo(elemento);
+        if (esVacia()) {
+            cabeza = rabo = nuevo;
+        } else {
+            rabo.siguiente = nuevo;
+            nuevo.anterior = rabo;
+            rabo = nuevo;
+        }
+        longitud++;
     }
 
     /**
@@ -119,7 +143,17 @@ public class Lista<T> implements Iterable<T> {
      *         <code>null</code>.
      */
     public void agregaInicio(T elemento) {
-        // Aquí va su código.
+        if (elemento == null) 
+            throw new IllegalArgumentException(MSJ_NULL_ELEM);
+        Nodo nuevo = new Nodo(elemento);
+        if (esVacia()) {
+            cabeza = rabo = nuevo;
+        } else {
+            cabeza.anterior = nuevo;
+            nuevo.siguiente = cabeza;
+            cabeza = nuevo;
+        }
+        longitud++;
     }
 
     /**
@@ -138,7 +172,26 @@ public class Lista<T> implements Iterable<T> {
      *         <code>null</code>.
      */
     public void inserta(int i, T elemento) {
-        // Aquí va su código.
+        if (elemento == null)
+            throw new IllegalArgumentException(MSJ_NULL_ELEM);
+        if (i <= 0) {
+            agregaInicio(elemento);
+            return;
+        }
+        if (i >= longitud) {
+            agregaFinal(elemento);
+            return;
+        }
+        Nodo nuevo = new Nodo(elemento);
+        Nodo temp = cabeza;
+        for (int j = 0; j < i - 1; j++) {
+            temp = temp.siguiente;
+        }
+        nuevo.anterior = temp;
+        nuevo.siguiente = temp.siguiente;
+        temp.siguiente.anterior = nuevo;
+        temp.siguiente = nuevo;
+        longitud++;
     }
 
     /**
@@ -147,7 +200,41 @@ public class Lista<T> implements Iterable<T> {
      * @param elemento el elemento a eliminar.
      */
     public void elimina(T elemento) {
-        // Aquí va su código.
+        if (esVacia() || elemento == null)
+            return;
+        Nodo elim = getNodo(elemento);
+        if (elim == null) 
+            return;
+        if (elim == cabeza) {
+            eliminaPrimero();
+            return;
+        } 
+        if (elim == rabo) {
+            eliminaUltimo();
+            return;
+        }
+        elim.anterior.siguiente = elim.siguiente;
+        elim.siguiente.anterior = elim.anterior;
+        longitud--;
+    }
+
+    /**
+     * Metodo auxiliar del metodo elimina.
+     * Regresa un Nodo que contega la primer ocurrencia del elemento recibido 
+     * como parametro, o <code>null</code> se no se encuentra.
+     * @param elemento el elemento a buscar.
+     * @return el Nodo con la primer ocurrencia del elemento.
+     */
+    private Nodo getNodo(T elemento) {
+        if (esVacia() || elemento == null) 
+            return null;
+        Nodo temp = cabeza;
+        while (temp != null) {
+            if (temp.elemento.equals(elemento))
+                return temp;
+            temp = temp.siguiente;
+        }
+        return null;
     }
 
     /**
@@ -156,7 +243,17 @@ public class Lista<T> implements Iterable<T> {
      * @throws NoSuchElementException si la lista es vacía.
      */
     public T eliminaPrimero() {
-        // Aquí va su código.
+        if (esVacia()) 
+            throw new NoSuchElementException(MSJ_EMPTY_LIST);
+        T elim = cabeza.elemento;
+        if (cabeza == rabo) {
+            cabeza = rabo = null;
+        } else {
+            cabeza = cabeza.siguiente;
+            cabeza.anterior = null;
+        }
+        longitud--;
+        return elim;
     }
 
     /**
@@ -165,7 +262,17 @@ public class Lista<T> implements Iterable<T> {
      * @throws NoSuchElementException si la lista es vacía.
      */
     public T eliminaUltimo() {
-        // Aquí va su código.
+        if (esVacia())
+            throw new NoSuchElementException(MSJ_EMPTY_LIST);
+        T elim = rabo.elemento;
+        if (cabeza == rabo) {
+            cabeza = rabo = null;
+        } else {
+            rabo = rabo.anterior;
+            rabo.siguiente = null;
+        }
+        longitud--;
+        return elim;
     }
 
     /**
@@ -175,7 +282,7 @@ public class Lista<T> implements Iterable<T> {
      *         <code>false</code> en otro caso.
      */
     public boolean contiene(T elemento) {
-        // Aquí va su código.
+        return getNodo(elemento) != null;
     }
 
     /**
@@ -183,7 +290,12 @@ public class Lista<T> implements Iterable<T> {
      * @return una nueva lista que es la reversa la que manda llamar el método.
      */
     public Lista<T> reversa() {
-        // Aquí va su código.
+        Lista<T> reversa = new Lista<>();
+        Iterator<T> it = iterator();
+        while (it.hasNext()) {
+            reversa.agregaInicio(it.next());
+        }
+        return reversa;
     }
 
     /**
@@ -192,14 +304,21 @@ public class Lista<T> implements Iterable<T> {
      * @return una copiad de la lista.
      */
     public Lista<T> copia() {
-        // Aquí va su código.
+        Lista<T> copia = new Lista<>();
+        Nodo temp = cabeza;
+        while (temp != null) {
+            copia.agregaFinal(temp.elemento);
+            temp = temp.siguiente;
+        }
+        return copia;
     }
 
     /**
      * Limpia la lista de elementos, dejándola vacía.
      */
     public void limpia() {
-        // Aquí va su código.
+        cabeza = rabo = null;
+        longitud = 0;
     }
 
     /**
@@ -208,7 +327,9 @@ public class Lista<T> implements Iterable<T> {
      * @throws NoSuchElementException si la lista es vacía.
      */
     public T getPrimero() {
-        // Aquí va su código.
+        if (esVacia())
+            throw new NoSuchElementException(MSJ_EMPTY_LIST);
+        return cabeza.elemento;
     }
 
     /**
@@ -217,7 +338,9 @@ public class Lista<T> implements Iterable<T> {
      * @throws NoSuchElementException si la lista es vacía.
      */
     public T getUltimo() {
-        // Aquí va su código.
+        if (esVacia())
+            throw new NoSuchElementException(MSJ_EMPTY_LIST);
+        return rabo.elemento;
     }
 
     /**
@@ -228,7 +351,13 @@ public class Lista<T> implements Iterable<T> {
      *         igual que el número de elementos en la lista.
      */
     public T get(int i) {
-        // Aquí va su código.
+        if (i < 0 || i >= longitud) 
+            throw new ExcepcionIndiceInvalido("Indice fuera de rango: [0, longitud).");
+        Nodo temp = cabeza;
+        for (int j = 0; j < i; j++) {
+            temp = temp.siguiente;
+        }
+        return temp.elemento;
     }
 
     /**
@@ -238,7 +367,15 @@ public class Lista<T> implements Iterable<T> {
      *         no está contenido en la lista.
      */
     public int indiceDe(T elemento) {
-        // Aquí va su código.
+        if (esVacia() || elemento == null) 
+            return -1;
+        Nodo temp = cabeza;
+        for (int i = 0; i < longitud; i++) {
+            if (temp.elemento.equals(elemento))
+                return i;
+            temp = temp.siguiente;
+        }
+        return -1;
     }
 
     /**
@@ -246,7 +383,15 @@ public class Lista<T> implements Iterable<T> {
      * @return una representación en cadena de la lista.
      */
     @Override public String toString() {
-        // Aquí va su código.
+        StringBuilder sb = new StringBuilder("[");
+        Nodo temp = cabeza; 
+        while (temp != null) {
+            sb.append(temp.elemento);
+            if (temp.siguiente != null)
+                sb.append(", ");
+            temp = temp.siguiente;
+        }
+        return sb.toString() + "]";
     }
 
     /**
@@ -259,7 +404,17 @@ public class Lista<T> implements Iterable<T> {
         if (objeto == null || getClass() != objeto.getClass())
             return false;
         @SuppressWarnings("unchecked") Lista<T> lista = (Lista<T>)objeto;
-        // Aquí va su código.
+        if (longitud != lista.longitud)
+            return false;
+        Nodo n1 = cabeza;
+        Nodo n2 = lista.cabeza;
+        for (int i = 0; i < longitud; i++) {
+            if (!n1.elemento.equals(n2.elemento))
+                return false;
+            n1 = n1.siguiente;
+            n2 = n2.siguiente;
+        }
+        return true;
     }
 
     /**
